feat: add web bundle as download option

roedoejet opened the pull request:
Here's a first stab at this. A few things:

- [x] The [web preview](https://readalong-studio.mothertongues.org/pr-preview/pr-241/) should work if you want to try it out (select "web bundle" from the download format options)
- [ ] ~I don't love how much logic is in the `demo.component.ts`~ I can live with this for now, but I think we're due for some refactoring
- [x] Maybe we should come up with a better basename than just `output` since if people are combining multiple readalongs they'll have to rename everything. timestamps?
- [x] The `assets` folder stuff needs to be improved. We have the `use-assets-folder` property, but the value is just a boolean. I think we should instead have that be a path that could get prepended to all images, because right now, the images are hardcoded in the `.readalong` file as being in an `assets` folder relative to the location of the ReadAlong, which is sure to break in many deployments. Update: see https://github.com/ReadAlongs/Web-Component/pull/243
- [x] looks like we need to run `prettier` on everything. I didn't do that here yet, to make sure the content changes are easily parsable.
- [x] Translations for french and spanish still have to be added

roedoejet committed:
feat: add web bundle as download option

joanise commented:
More comments, now that I have tested this feature.

File names:
 - `readalong.zip` would be better than `output.zip`, more self-documenting.
 - Ditto the files in `assets/` in the archive, `assets/readalong.{mp3,readalong}` would be better than `assets/output.*`.
 - `output.txt` is confusing, since it's really the input text. `readalong.txt` would be OK, or maybe `input.txt`.

The `README.txt` file should be outdented against the left margin. My guess is you have it indented probably because of the file where you're defining its contents, but I don't like the result, it ought to start on the first line (remove the blank line) and have no indents. It should also be wrapped at <=79 characters. I know we don't do that anymore in code, but here when I double clicked on the file the display was awful at 80 chars wide.
![image](https://github.com/ReadAlongs/Web-Component/assets/33264708/10f5854b-0510-4fd3-8668-c4c60870785c)
I would also be OK with no manual wrapping at all, letting the auto-wrapping do its thing, but pre-wrapped at 107 yields poor results if the window is even slightly narrower. Between not pre-wrapped at all, and pre-wrapped at 79, I can't decide what I prefer: not pre-wrapped adjusts to whatever window size the user has, 79 is old machine friendly, though maybe not mobile friendly... You pick!
Key is, I'd like the default display to be easy to read on all OSes, machine ages, and user defaults.

joanise commented:
I tested unzipping the file on my machine and running `python -m http.server`, and that works nicely.

One thing, though, this is a zip bomb, with no inner folder. Would it make sense to put all the contents in a `readalong/` folder inside the zip file? Or is it just because I'm a developer that I'm allergic to tar bombs?

joanise commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -35,10 +44,30 @@ export class DemoComponent implements OnDestroy, OnInit {
+  readmeFile = new Blob(
+    [
+      `
+    Web Development Guide
```
Web Deployment Guide?

joanise commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -27,6 +35,7 @@ export class DemoComponent implements OnDestroy, OnInit {
   };
   outputFormats = [
     { value: "html", display: $localize`Offline HTML` },
+    { value: "zip", display: $localize`Web Bundle` },
```
Zipped Web Bundle?
That would give me "Fichiers Web zippés" and "Archivos web comprimidos".
I find just "Web Bundle" insufficiently clear.

joanise replied:
I did some research on "zippés" vs "comprimés" for the French translation, and Google tells me the former is much more common than the latter, and that matches my intuition, but Spanish doesn't seem to have that, it would have to be "Archivos web comprimidos zip" if we wanted to be precise, but I think that's unnecessary, adding more verbosity with no additional clarity, whereas the French "zippés" is shorter and more common, thus a no brainer.

dhdaines replied:
compressés: https://vitrinelinguistique.oqlf.gouv.qc.ca/fiche-gdt/fiche/8399261/compresser

dhdaines replied:
mais même selon l'OQLF "zippé" est acceptable dans certains contextes ;-)

joanise replied:
I'm not going to be picky, but my first choice is "zippé" because we're creating an actual .zip file.

Does anyone have an opinion on the English, though, whether we should add "Zipped" in front?

roedoejet replied:
I think "zipped" is redundant - Bundle implies that it's bundled/zipped which is sufficient

joanise commented:
By the way, we already applied prettier:

```
$ npx prettier -c packages
Checking formatting...
All matched files use Prettier code style!
```

the generated `index.html` file is not prettier'd, though, that's true.

dhdaines commented:
> The `README.txt` file should be outdented against the left margin. My guess is you have it indented probably because of the file where you're defining its contents, but I don't like the result, it ought to start on the first line (remove the blank line) and have no indents. It should also be wrapped at <=79 characters. I know we don't do that anymore in code, but here when I double clicked on the file the display was awful at 80 chars wide.

`black` wraps at "90-ish" so we do this in code, more or less.  I think it is reasonable to wrap at 80 (if you use Emacs it is as simple as typing `Esc-q`) as this makes the text easier to read in general (there's a reason why LaTeX makes wide margins).  If you don't wrap lines then not everbody's terminal/browser/whatever is going to actually word wrap correctly in any case.

dhdaines commented:
> One thing, though, this is a zip bomb, with no inner folder. Would it make sense to put all the contents in a `readalong/` folder inside the zip file? Or is it just because I'm a developer that I'm allergic to tar bombs?

Yes, I think this is a good idea - MacOS or Windows are going to unpack it in a new folder anyway, but this way nobody gets any nasty surprises.

dhdaines commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -189,6 +235,74 @@ export class DemoComponent implements OnDestroy, OnInit {
+            </body>
+
+            <!-- The last step needed is to import the package -->
+            <script type="module" src='https://unpkg.com/@readalongs/web-component@^1.0.0/dist/web-component/web-component.esm.js'></script>
```
Should we instead make sure this is the same version as the web application?  I think we can get it from the environment during build

dhdaines commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -189,6 +235,74 @@ export class DemoComponent implements OnDestroy, OnInit {
+      // - write zip
+      zipFile
+        .generateAsync({ type: "blob" })
+        .then((content) => saveAs(content, `${basename}.zip`));
```
Could this promise be rejected?  We should handle possible errors here.

dhdaines commented on packages/studio-web/src/app/soundswallower.service.spec.ts:
```diff
@@ -583,12 +588,19 @@ AFwAWQBbAGEAagBvAHAAcABkAGIAaABhAGcAagBwAGkAVwBfAA==`;
+  let httpTestingController: HttpTestingController;
   beforeEach(() => {
-    TestBed.configureTestingModule({});
+    httpClientSpy = jasmine.createSpyObj("HttpClient", ["get"]);
```
Jest testing seems to work with nx and angular now... when I do the angular 17 migration I will eradicate Jasmine for good

dhdaines commented on packages/studio-web/src/app/upload.service.ts:
```diff
@@ -0,0 +1,11 @@
+})
+export class UploadService {
+  $currentAudio = new BehaviorSubject<Blob | File | null>(null);
+  $currentText = new BehaviorSubject<Blob | string | null>(null);
```
What does the convention say to do with the `$` anyway?  Before or after the name?  Does it depend if it's a `Subject` or an `Observable`?

roedoejet replied:
I thought a $ prefix was subject and a $ suffix was observable. But it looks like Angular only explicitly documents the use of $ as a suffix for observables.

dhdaines replied:
Hmm - now that I think of it, `Subject`s are also `Observable`s, so that makes sense.

roedoejet replied:
true, but I kind of like having $ prefix since it lets you know that you can call `next` on it etc.. but I don't feel super strongly about it. I just changed it to `$textInput` but I can change it if you prefer

dhdaines replied:
Makes sense!  I agree it's useful to have a $ on it so we know what it is...

dhdaines commented on packages/studio-web/src/app/upload/upload.component.ts:
```diff
@@ -83,7 +84,7 @@ export class UploadComponent implements OnDestroy, OnInit {
     text: "edit",
   };
-  textInput: string = "";
+  textInput = new BehaviorSubject<string>("");
```
Since this is an `rxjs` thingy it should have a `$` ... somewhere ... in the name to be consistent.

dhdaines commented on packages/studio-web/src/app/upload/upload.component.ts:
```diff
@@ -92,11 +93,23 @@ export class UploadComponent implements OnDestroy, OnInit {
     private dialog: MatDialog
-  ) {}
+  ) {
+    this.audioControl.valueChanges
```
Likewise with all of these rxjs creatures

dhdaines reviewed (changes requested):
Thank you!  Looks good, mostly style comments.

roedoejet committed:
fix: include review-recommended fixes

	- Update $ style for rxjs variables
	- Add inner folder for zip
	- Use timestamps for filenames
	- Remove indentation for readme
	- Added explicit version for unpkg import
	- Added error handling for zip download promise

roedoejet committed:
fix(assets)!: get rid of useAssetsFolder boolean and replace with imageAssetsFolder string

BREAKING CHANGE: useAssetsFolder is no longer valid.

The equivalent is imageAssetsFolder='assets/' which is the default.

joanise commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -189,6 +237,88 @@ export class DemoComponent implements OnDestroy, OnInit {
+      // Create inner folder
+      const innerFolder = zipFile.folder("readalong");
+      const assetsFolder = innerFolder?.folder("assets");
+      const timestamp = Date.now();
```
This gives us a timestamp in seconds since epoch, e.g., `readalong-1710864009240`.
Can we do yyyymmddhhmmss instead? It's just one digit longer but it's human readable.

joanise commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -35,10 +44,31 @@ export class DemoComponent implements OnDestroy, OnInit {
+
+Your index.html file demonstrates the snippet and imports needed to host the ReadAlong on your server.
+
+Please host all assets on your server, include the font and package imports defined in the index.html in your website's imports, and include the corresponding <readalong> snippet everywhere you would like your ReadAlong to be displayed.
```
That should say `<read-along>`, not `<readalong>`. It look me a while to notice this, but it's been nagging me, something felt wrong but I didn't quite know what... :)

roedoejet committed:
fix: use human readable timestamp

roedoejet committed:
docs(i18n): add es and fr translations

joanise committed:
docs(i18n): translate the new message "download failed"

joanise reviewed (approved):
Looks good to me now.

dhdaines reviewed (approved):
looks good! sorry for the delay in approving!

deltork commented:
can the readme contain section about deployment on wordpress, with the apporpiate snippet :
   ` [wp_read_along_web_app_loader  version='^1.2.0']
    <read-along...></readalong>
    [/wp_read_along_web_app_loader]`

joanise commented:
> can the readme contain section about deployment on wordpress, with the apporpiate snippet :
> ` [wp_read_along_web_app_loader version='^1.2.0'] <read-along...></readalong> [/wp_read_along_web_app_loader]`

Absolutely, just add that in your PR.

deltork commented:
> > can the readme contain section about deployment on wordpress, with the apporpiate snippet :
> > ` [wp_read_along_web_app_loader version='^1.2.0'] <read-along...></readalong> [/wp_read_along_web_app_loader]`
> 
> Absolutely, just add that in your PR.

The readme is only in this branch. I will have to add the instructions after this branch is merged

deltork commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -189,6 +237,91 @@ export class DemoComponent implements OnDestroy, OnInit {
+        .toISOString()
+        .replace(/[^0-9]/g, "")
+        .slice(0, -3);
+      const basename = `readalong-${timestamp}`;
```
When people provide a title, adding the title to the filename will help them find it after download, especially if they do multiple books.
`` const basename = (this.slots.title?this.slots.title.replace(/\s/g,"_").toLowerCase():'readalong')+`-${timestamp}`; ``

roedoejet committed:
fix: check that assets folder ends with /

roedoejet committed:
feat: use slot title as file basename in bundle

joanise commented on packages/studio-web/src/app/demo/demo.component.ts:
```diff
@@ -160,6 +209,13 @@ export class DemoComponent implements OnDestroy, OnInit {
+        .replace(/[^0-9]/g, "")
+        .slice(0, -3);
+      const basename =
+        (this.slots.title ? slugify(this.slots.title, 15) : "readalong") +
```
This is great, except it's too anglocentric: if the title is ᐃᓄᒃᑎᑐᑦ, the slug is empty and you get `-20240402212031.zip`, and the .txt file inside is similarly limited. Similarly "Éric" becomes "ric", but the Inuktitut example is a lot more dramatic. Can we slugify the text-unidecode output, maybe?

roedoejet replied:
thanks for this @joanise - I've updated it to use unidecode

roedoejet committed:
feat: use unidecode to create filenames

roedoejet committed:
feat: bump to v1.2.0

roedoejet committed:
chore: bump single file bundle version
